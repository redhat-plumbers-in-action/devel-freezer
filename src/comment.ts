import { Context } from 'probot';
import { debug } from 'console';

import { events } from './events';
import { getMetadata, setMetadata } from './metadata';
import { warning } from '@actions/core';

const develFreezerCommentID = 'devel-freezer-comment-id';
const develFreezerFreezingTag = 'devel-freezer-freezing-tag';

export async function publishComment(
  content: string,
  tag: string,
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
) {
  const commentID = await getCommentID(context);

  if (commentID) {
    updateComment(commentID, content, context);
    return;
  }

  const commentPayload = (await createComment(content, context))?.data;

  if (!commentPayload) {
    warning(`Failed to create comment.`);
    return;
  }

  // Set metadata
  await setCommentID(commentPayload.id, context);
  // TODO: Probably dont work when updating comment!!!
  await setFeezingTag(tag, context);
}

async function getCommentID(
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
): Promise<number | undefined> {
  return getMetadata(develFreezerCommentID, context as unknown as Context);
}

async function setCommentID(
  id: number,
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
) {
  debug(`id: '${id}'`);

  await setMetadata(
    develFreezerCommentID,
    id.toString(),
    context as unknown as Context
  );
}

async function setFeezingTag(
  tag: string,
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
) {
  debug(`tag: '${tag}'`);

  await setMetadata(
    develFreezerFreezingTag,
    tag.toString(),
    context as unknown as Context
  );
}

function createComment(
  body: string,
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
) {
  if (!body || body === '') return;

  return context.octokit.issues.createComment(
    context.issue({
      body,
    })
  );
}

async function updateComment(
  commentID: number,
  body: string,
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
) {
  return context.octokit.issues.updateComment(
    context.issue({
      comment_id: commentID,
      body,
    })
  );
}
