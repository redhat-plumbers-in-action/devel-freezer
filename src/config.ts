import {
  ArrayMinSize,
  IsArray,
  IsString,
  MinLength,
  validate,
  ValidateNested,
} from 'class-validator';
import { Context } from 'probot';

import { events } from './events';
import { ValidationFeedback } from './validation-feedback';

import { TConfigObject, TFeedback, TPolicyItem } from './types.d';

export class Config {
  @IsArray()
  @ValidateNested({ each: true })
  @ArrayMinSize(1)
  private _policy: PolicyItem[];

  constructor(config: TConfigObject) {
    this._policy = Array.isArray(config?.policy)
      ? config.policy.map(item => new PolicyItem(item))
      : [];
  }

  get policy() {
    return this._policy;
  }

  static async getConfig(
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    const retrievedConfig = await context.config<TConfigObject>(
      'development-freeze.yml'
    );

    if (Config.isConfigEmpty(retrievedConfig)) {
      throw new Error(
        `Missing configuration. Please setup 'devel-freezer' Action using 'development-freeze.yml' file.`
      );
    }

    const config = new this(retrievedConfig as TConfigObject);

    return config;
  }

  static isConfigEmpty(config: TConfigObject | null | unknown) {
    return config === null;
  }

  static async validate(instance: Config) {
    const validationResult = await validate(instance, {
      whitelist: true,
      forbidNonWhitelisted: true,
    });

    const results = validationResult.map(error => {
      return ValidationFeedback.composeFeedbackObject(error);
    });

    return results;
  }
}

class PolicyItem {
  @IsString({ each: true })
  @MinLength(1, { each: true })
  private _tags: string[];

  @ValidateNested()
  private _feedback: Feedback;

  constructor(item: TPolicyItem) {
    this._tags = item.tags;
    this._feedback = new Feedback(item.feedback);
  }

  get tags() {
    return this._tags;
  }

  get feedback() {
    return this._feedback;
  }
}

class Feedback {
  @IsString()
  @MinLength(1)
  private _frozenState: string;

  @IsString()
  @MinLength(1)
  private _unFreezeState: string;

  constructor(feedback: TFeedback) {
    this._frozenState = feedback['frozen-state'];
    this._unFreezeState = feedback['unfreeze-state'];
  }

  get frozenState() {
    return this._frozenState;
  }

  get unFreezeState() {
    return this._unFreezeState;
  }
}
