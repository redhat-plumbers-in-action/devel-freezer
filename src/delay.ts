import { debug, warning } from '@actions/core';
import { setTimeout } from 'timers/promises';

export async function delay(seconds: number) {
  warning(`Delaying the action for ${seconds} seconds.`);

  await setTimeout(seconds * 1000);

  debug('Delay is over.');
}
