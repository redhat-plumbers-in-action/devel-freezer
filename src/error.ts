export class FreezerError extends Error {
  constructor(
    message: string,
    readonly code?: number
  ) {
    super(message);
  }
}

export function raise(error: string, code?: number): never {
  throw new FreezerError(error, code);
}
