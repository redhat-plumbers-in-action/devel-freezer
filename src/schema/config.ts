import { z } from 'zod';

export const policySchema = z.object({
  tags: z.array(z.string().min(1)),
  labels: z
    .object({
      allow: z.array(z.string().min(1)),
    })
    .optional(),
  feedback: z.object({
    'frozen-state': z.string().min(1),
    'unfreeze-state': z.string().min(1),
  }),
});
export type PolicySchema = z.infer<typeof policySchema>;

export const configSchema = z.object({
  policy: z.array(policySchema).min(1),
});

export type ConfigType = z.infer<typeof configSchema>;
