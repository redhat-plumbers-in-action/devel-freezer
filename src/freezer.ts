import { Context } from 'probot';

import { events } from './events';
import { publishComment } from './comment';

export async function freezePullRequest(
  comment: string,
  tag: string,
  context: {
    [K in keyof typeof events]: Context<typeof events[K][number]>;
  }[keyof typeof events]
) {
  await publishComment(comment, tag, context);
}
