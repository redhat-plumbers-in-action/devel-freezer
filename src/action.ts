import { debug, warning, getInput } from '@actions/core';

import { Config } from './config';
import { CustomOctokit } from './octokit';
import { PullRequest } from './pull-request';
import { Tag } from './tag';
import { delay } from './delay';

import { inputDelaySchema } from './schema/inputs';

async function action(octokit: CustomOctokit): Promise<void> {
  const delayParsed = inputDelaySchema.safeParse(getInput('delay'));
  const delaySeconds = delayParsed.success ? delayParsed.data : 0;

  if (delaySeconds > 0) {
    await delay(delaySeconds);
  }

  const config = await Config.getConfig(octokit);

  const tag = new Tag(await Tag.getLatestTag());

  if (!tag.latest) {
    warning(`Repository doesn't have any tags or releases published.`);
    return;
  }

  debug(`Latest tag is: '${tag.latest}'`);

  const pullRequest = await PullRequest.getPullRequest(
    +getInput('pr-number'),
    octokit
  );

  for (const policyItem of config.policy) {
    if (!tag.isFreezed(policyItem.tags)) {
      continue;
    }

    await pullRequest.freeze(policyItem.feedback['frozen-state'], tag.latest);
    return;
  }

  if (!pullRequest.isFreezed()) {
    return;
  }

  for (const policyItem of config.policy) {
    if (!pullRequest.isTagPolicyCompliant(policyItem.tags)) {
      continue;
    }

    await pullRequest.unfreeze(policyItem.feedback['unfreeze-state']);
    return;
  }

  debug(
    `The latest tag doesn't match the requirements for a development freeze.`
  );
}

export default action;
