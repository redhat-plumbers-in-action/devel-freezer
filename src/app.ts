import { debug, error, warning } from '@actions/core';
import { Context, Probot } from 'probot';
import { validateOrReject } from 'class-validator';

import { events } from './events';
import { freezePullRequest } from './freezer';
import { getLatestTag, isTagFreezed } from './release';
import { Config } from './config';

const app = (probot: Probot) => {
  probot.on(
    events.pull_request,
    async (context: Context<typeof events.pull_request[number]>) => {
      const config = await Config.getConfig(context);

      await validateOrReject(config);

      if (!config) {
        error(
          `Missing configuration. Please setup 'devel-freezer' action using 'development-freeze.yml' file.`
        );
        return;
      }

      const latestTag = await getLatestTag(context);

      if (!latestTag) {
        warning(`Repository doesn't have any tags or releases published.`);
        return;
      }

      debug(`Latest tag is: '${latestTag}'`);

      for (const item of config.policy) {
        if (!isTagFreezed(latestTag, item.tags)) {
          continue;
        }

        await freezePullRequest(item.feedback.freezedState, latestTag, context);
        return;
      }

      // TODO: If PR is changed and contains metadata from devel-freezer, update comment
      // ? It requires knowledge about old freezing tag ...

      // get metadata if any
      // if metadata check tag and get unfreezing comment (for loop policy)
      // update metadata and comment

      debug(
        `The latest tag doesn't match the requirements for a development freeze.`
      );
    }
  );
};

export default app;
