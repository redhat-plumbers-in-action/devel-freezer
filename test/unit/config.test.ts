import { describe, it, expect, beforeEach, test } from 'vitest';

import { Config } from '../../src/config';

import {
  configContextFixture,
  IConfigTestContext,
} from './fixtures/config.fixture';

describe('Config Object', () => {
  beforeEach<IConfigTestContext>(context => {
    context.configs = configContextFixture.configs;
    context.invalid = configContextFixture.invalid;
  });

  it<IConfigTestContext>('can be instantiated', context => {
    context.configs.map(item => expect(item).toBeDefined());
  });

  test<IConfigTestContext>('get policy()', context => {
    context.configs.map(configItem => {
      expect(configItem.policy).toMatchSnapshot();

      configItem.policy.map(policyItem => {
        expect(policyItem.tags).toMatchSnapshot();
        expect(policyItem.feedback).toMatchSnapshot();
        expect(policyItem.feedback.frozenState).toMatchSnapshot();
        expect(policyItem.feedback.unFreezeState).toMatchSnapshot();
      });
    });
  });

  it<IConfigTestContext>('is valid', async context => {
    context.configs.map(async item =>
      expect(Config.validate(item)).resolves.toMatchObject([])
    );
  });

  it<IConfigTestContext>('is invalid', async context => {
    context.invalid.map(async item =>
      expect(Config.validate(item)).resolves.toMatchSnapshot()
    );
  });

  test<IConfigTestContext>('is config empty', context => {
    context.configs.map(async item =>
      expect(Config.isConfigEmpty(item)).toEqual(false)
    );

    context.invalid.map(async item =>
      expect(Config.isConfigEmpty(item)).toEqual(false)
    );

    expect(Config.isConfigEmpty(null)).toEqual(true);
  });
});
