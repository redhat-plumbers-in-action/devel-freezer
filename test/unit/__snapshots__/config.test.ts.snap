// Vitest Snapshot v1

exports[`Config Object > get policy() 1`] = `
[
  PolicyItem {
    "_feedback": Feedback {
      "_frozenState": "This is No-No",
      "_unFreezeState": "This is Yes-Yes",
    },
    "_tags": [
      "alpha",
      "beta",
    ],
  },
]
`;

exports[`Config Object > is invalid 1`] = `
[
  {
    "notes": {
      "arrayMinSize": "_policy must contain at least 1 elements",
    },
    "property": "_policy",
    "value": [],
  },
]
`;

exports[`Config Object > is invalid 2`] = `
[
  {
    "notes": {
      "arrayMinSize": "_policy must contain at least 1 elements",
    },
    "property": "_policy",
    "value": [],
  },
]
`;

exports[`Config Object > is invalid 3`] = `
[
  {
    "notes": {
      "arrayMinSize": "_policy must contain at least 1 elements",
    },
    "property": "_policy",
    "value": [],
  },
]
`;

exports[`Config Object > is invalid 4`] = `
[
  {
    "notes": {
      "isString": "_frozenState must be a string",
      "minLength": "_frozenState must be longer than or equal to 1 characters",
    },
    "property": "_policy.0._feedback._frozenState,_unFreezeState",
    "value": undefined,
  },
]
`;
